src/async-multi-partition.mjs

```diff
@@ -0,0 +1,56 @@
+import { Queue, fakeQueue } from './internal/queues'
+import { asyncIterableCurry } from './internal/async-iterable'
+
+function * asyncMultiPartition (func, iter) {
+  const queues = []
+  let maxQueues = Infinity
+  let queueNumber = 0
+  const iterator = iter[Symbol.asyncIterator]()
+  let exhausted = 0
+
+  async function returnIterator () {
+    if (exhausted === maxQueues) {
+      if (typeof iterator.return === 'function') iterator.return()
+    }
+  }
+
+  async function * part (queueId) {
+    yield 'ready'
+    queues[queueId] = queues[queueId] || new Queue()
+    try {
+      while (true) {
+        while (!queues[queueId].isEmpty()) {
+          yield queues[queueId].shift()
+        }
+
+        const { value, done } = await iterator.next()
+        if (done) break
+
+        const chosen = await func(value)
+        if (chosen < maxQueues) { // throw away item if queue doesn't exist
+          queues[chosen] = queues[chosen] || new Queue()
+          queues[chosen].push(value)
+        }
+      }
+    } finally {
+      queues[queueId] = fakeQueue // /dev/null
+      exhausted++
+      await returnIterator()
+    }
+  }
+
+  try {
+    while (true) {
+      const iter = part(queueNumber++)
+      // this ensure to call the finally clause of iter
+      iter.next()
+      yield iter
+    }
+  } finally {
+    // I trim the number of queues
+    queues.length = maxQueues = queueNumber
```

Comment:
Maybe this is what causes [the issue that you talked about](https://github.com/sithmel/iter-tools/pull/199#issuecomment-466798162).

Reply:
no, the cause was very subtle. But it is now fixed https://github.com/sithmel/iter-tools/pull/199/commits/895572f3cd7486f0cf67b10ea6a26acbba86212d#diff-bc98b2acad88eb39be94dcd71e05d22bR27